docs/developer/solutions/archive.rst

```diff
@@ -0,0 +1,177 @@
+   .
+   ├── images
+   │   └── some_image_name
+   │       └── 1.0.1
+   │           ├── <layer_digest>
+   │           ├── manifest.json
+   │           └── version
+   ├── operator
+   │   ├── crd
+   │   │   └── some_crd_name.yaml
+   │   └── deployment.yaml
+   ├── product.txt
+   └── ui
+       └── deployment.yaml
+
+.. _solution-archive-product-info:
+
+Product information
+-------------------
+
+General product information about the packaged Solution must be stored in the
+``product.txt`` file, stored at the archive root.
+
+It must respect the following format::
+
+   NAME=Example
+   VERSION=1.0.0-dev
+   REQUIRE_METALK8S=">=2.0"
+
+.. note::
+
+   If a Solution can require specific versions of MetalK8s on which to be
+   deployed, requiring specific services (and their respective versions) to be
+   shipped with MetalK8s (e.g. Prometheus/Grafana) is not yet feasible.
+   It will probably be handled in the Operator declaration, maybe using a CR.
+
+.. todo::
+
+   Consider if more keys are needed, e.g. author/maintainer, Git reference
+
+
+.. _solution-archive-images:
+
+OCI images
+----------
+
+MetalK8s exposes container images in the OCI_ format through a static
+read-only registry. This registry is built with nginx_, and relies on having
+a specific layout of image layers to then replicate the necessary parts of the
+Registry API that CRI clients (such as ``containerd`` or ``cri-o``) rely on.
+
+Using skopeo_, you can save images as a directory of layers::
+
+   $ mkdir images/my_image
+   $ # from your local Docker daemon
+   $ skopeo copy --format v2s2 --dest-compress docker-daemon:my_image:1.0.0 dir:images/my_image/1.0.0
+   $ # from Docker Hub
+   $ skopeo copy --format v2s2 --dest-compress docker://docker.io/example/my_image:1.0.0 dir:images/my_image/1.0.0
+
+Your ``images`` directory should now resemble this::
+
+   images
+   └── my_image
+       └── 1.0.0
+           ├── 53071b97a88426d4db86d0e8436ac5c869124d2c414caf4c9e4a4e48769c7f37
+           ├── 64f5d945efcc0f39ab11b3cd4ba403cc9fefe1fa3613123ca016cf3708e8cafb
+           ├── manifest.json
+           └── version
+
+Once all your images were stored this way, you can de-duplicate layers using
+hardlinks, using the tool hardlink_::
+
+   $ hardlink -c images
+
+A detailed procedure for generating the expected layout is available at
+`NicolasT/static-container-registry`_. You can use the script provided there
+to generate your own NGINX configuration and test this static registry for
+yourself. MetalK8s will generate its own configuration when importing the
+Solution archive.
+
+Each archive will be exposed as a single repository, where the name will be
+computed as ``<NAME>-<VERSION>`` from :ref:`solution-archive-product-info`.
+
+.. warning::
+
+   Operator Deployments should not rely on this naming pattern for finding
+   the images for their resources. Instead, the full repository prefix will be
+   stored in a ``ConfigMap``, that Deployments will be able to expose as
```

Comment:
Because it needs to be readable somewhere that is not Salt, if users want to know where their images are stored. I thought `ConfigMap`s could help provide a consistent storage destination such that mounting its value as an environment variable in Operator Deployments is merely about knowing the `ConfigMap` name and the key for this Solution.

Or do we know the naming of repositories will be stable/predictable and only need to pass the registry address (in which case we probably just need to read the right `Service` object)?

Reply:
Fair enough, though I don't think solutions should be aware of this `ConfigMap`, or refer to it in their deployment manifest. That's cumbersome. We can provide or inject that value when rendering the manifest directly.